src/arithmetic/canonical.cc

```diff
@@ -740,12 +740,128 @@ T Simplify_(T a, Map<Var, Range> vrange) {
 }
 
 
+/*!
+ * \brief Simplify just the combiner of the given reduce node.
+ *
+ *  This function applies Simplify to the components of the top reduction's
+ *  combiner, but not to the source or condition of the reduction.
+ *  It also removes all components which are not used to
+ *  compute the resulting value (the value_index-th value).
+ *
+ *  If \p expr is not a reduction node, it is left unchanged.
+ *
+ * \param expr The expression to be simplifed.
+ * \return Simplified expression.
+ */
+Expr SimplifyCombiner(const Expr& expr, const Map<Var, Range>& vrange = Map<Var, Range>()) {
+  const Reduce* op = expr.as<Reduce>();
+  if (!op)
+    return expr;
+
+  // First simplify the results
+  Array<Expr> simplified_result;
+  for (const auto& res : op->combiner->result)
+    simplified_result.push_back(Simplify(res, vrange));
+
+  // Which components to keep
+  std::vector<int> used(op->combiner->result.size(), false);
+
+  // This function recursively marks the used components starting from
+  // the index idx
+  std::function<void(int)> mark_used;
+  mark_used = [&used, &simplified_result, op, &mark_used](size_t idx) {
+    // if the idx-th component was marked as used before, do nothing
+    if (used[idx]) return;
+    used[idx] = true;
+
+    // check if the idx-th result expr uses some lhs or rhs variables
+    // and recursively mark the corresponding components
+    for (size_t i = 0; i < simplified_result.size(); ++i)
+      if (!used[i]) {
+        if (ExprUseVar(simplified_result[idx], op->combiner->lhs[i]) ||
+            ExprUseVar(simplified_result[idx], op->combiner->rhs[i]))
+          mark_used(i);
+      }
+  };
+
+  // mark all used components starting from the value_index
+  mark_used(op->value_index);
+
+  int new_value_index = op->value_index;
+  Array<Expr> new_result;
+  Array<Expr> new_identity;
+  Array<Var> new_lhs;
+  Array<Var> new_rhs;
+  Array<Expr> new_source;
+
+  // new stuff is old stuff which is used
+  for (size_t i = 0; i < used.size(); ++i) {
+    if (used[i]) {
```

Comment:
In addition to this, I think we should also keep `simplified_result[i]` when `HasSideEffect` returns `true`(highly unlikely, but for correctness). What do you think?

Reply:
Currently `HasSideEffect` is very conservative, e.g. it returns true if there is a tensor indexing, i.e. almost always. Moreover, it is never checked during arithmetic simplifications, so I guess it is implicitly assumed that there are no side effects at this level (which I think should be assumed).

Reply:
I guess so. `Simplify` simply calls into `HalideIR::simplify`(which doesn't seem to care about side effect), but `CanonicalSimplify` relies on side effect check. We might still need to confirm the assumption, if indeed we can assume that, let's add a comment. @tqchen, could you chime in?

Reply:
Oh, I see, `CanonicalSimplify` does indeed check for side effects (it just doesn't call the `HasSideEffect` function, so I didn't notice), so I was wrong and there is no such implicit assumption. Checking for side effects with the current definition of a side effect makes unused component elimination in combiner simplification quite useless though, so I don't know. Also eager const folding doesn't check for side effects, so it's a bit inconsistent already. I have strong doubts that we really need any guarantees about operations with side effects, but I may be wrong, so I would like too hear some argumentation behind including these checks.

Reply:
Eager constant folder won't fold functions, so it is still consistent with side effect checks.

Reply:
@sgrechanik-h https://github.com/dmlc/tvm/commit/64e5681ddd8dc0f8d4e45e912e6191e537632409 should now consider array index expression as side effect free, could you add side effect check here now?

Reply:
Yes, sure